lib/posix-socket/include/uk/socket_driver.h

```diff
@@ -0,0 +1,633 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __UK_SOCKET_DRIVER_H__
+#define __UK_SOCKET_DRIVER_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <uk/essentials.h>
+#include <uk/assert.h>
+#include <uk/alloc.h>
+#include <uk/socket.h>
+#include <uk/errptr.h>
+
+#include <sys/socket.h>
+#include <sys/uio.h>
+#include <errno.h>
+
+struct posix_socket_ops;
+struct posix_socket_file;
+
+struct eventpoll_cb;
+
+/**
+ * The POSIX socket driver defines the operations to be used for the
+ * specified AF family as well as the memory allocator.
+ */
+struct posix_socket_driver {
+	/** The AF family ID */
+	const int family;
+	/** Name of the driver library */
+	const char *libname;
+	/** The interfaces for this socket */
+	const struct posix_socket_ops *ops;
+	/** The memory allocator to be used for this socket driver */
+	struct uk_alloc *allocator;
+	/** Private data for this socket driver */
+	void *private;
+};
+
+/**
+ * The initialization function called for this socket family. It's here that
+ * additional configuration for the driver can be made after it has been
+ * registered. For instance, an alternative memory allocator can be provided.
+ *
+ * @param d The socket driver
+ *
+ * @return 0 on success, -errno code otherwise
+ */
+typedef int (*posix_socket_driver_init_func_t)(struct posix_socket_driver *d);
+
+/**
+ * Create a connection on a socket.
+ *
+ * @param driver The socket driver
+ * @param family Specifies a communication family domain and thus driver
+ * @param type Specifies communication semantics
+ * @param protocol Specifies a particular protocol to be used with the socket
+ *
+ * @return NULL or a valid pointer to driver-specific data on success,
+ *    ERR2PTR(-errno) otherwise
+ */
+typedef void *(*posix_socket_create_func_t)(struct posix_socket_driver *d,
+		int family, int type, int protocol);
+
+/**
+ * Accept a connection on a socket.
+ *
+ * @param sock Reference to the socket.
+ * @param addr The address of the peer socket.
+ * @param addr_len Specifies the size, in bytes, of the address structure
+ *    pointed to by addr
+ *
+ * @return NULL or a valid pointer to driver-specific data on success,
+ *    ERR2PTR(-errno) otherwise
+ */
+typedef void *(*posix_socket_accept_func_t)(struct posix_socket_file *sock,
```

Comment:
The `posix_socket_accept4_func_t` type is missing. Is this intentional?

Reply:
@razvand Yes, this was intentional. This is the driver interface. It is thus not bound to the POSIX standard but only resembles the interface for simplicity. You may notice that there is no `recv` and `send` either. Instead both are realized by expecting the driver to implement the more powerful `recvfrom` and `sendto` calls and `posix-socket` internally calls these methods.

I do not want a driver having to implement `accept` and `accept4`, so I changed the interface so that a driver needs to provide `accept4` only.